### PR TITLE
[Bug-Fix]:properties `current_epoch` and `global_step` between model and trainer same always 

### PR DESCRIPTION
**Pull request by nrupatunga**
What does this PR do?
- This PR makes sure that properties `current_epoch` and `global_step` shared between the lightning module and the trainer remain the same always. 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## Did you have fun?
Yes I had fun learning from Adrian 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -122,6 +116,14 @@ def __init__(self, *args, **kwargs):
     def example_input_array(self) -> Any:
         return self._example_input_array
 
+    @property
+    def current_epoch(self):
+        return self.trainer.current_epoch if self.trainer else 0
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:121]: - what happens if the module is loaded from checkpoint? 
- should these be -1 if `self.trainer` isn't set? that could indicate that this is currently outside of training

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -122,6 +116,14 @@ def __init__(self, *args, **kwargs):
     def example_input_array(self) -> Any:
         return self._example_input_array
 
+    @property
+    def current_epoch(self):
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:120]:
```suggestion
    def current_epoch(self) -> int:
```

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -122,6 +116,14 @@ def __init__(self, *args, **kwargs):
...
+        return self.trainer.current_epoch if self.trainer else 0
+
+    @property
+    def global_step(self):
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:124]:
```suggestion
    def global_step(self) -> int:
```

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (nrupatunga) [reply pytorch_lightning/core/lightning.py:121]: This could be a nice addition. 
- have set it to `0` as previously it was `self.current_epoch = 0` in `__init__`

**Comment by ydcjeff** [comment]
Can you please re-add those properties comments?

File: tests/models/test_restore.py [hunk 4]
```diff
@@ -10,11 +10,52 @@
...
 import tests.base.develop_utils as tutils
-from pytorch_lightning import Trainer
+from pytorch_lightning import Trainer, Callback
 from pytorch_lightning.callbacks import ModelCheckpoint
 from tests.base import EvalModelTemplate, GenericEvalModelTemplate
 
 
+class ModelTrainerPropertyParity(Callback):
+
+    def check_properties(self, trainer, pl_module):
```
Comment (Borda) [inline tests/models/test_restore.py:20]:
```suggestion
    def _check_properties(self, trainer, pl_module):
```

**Review by Borda — COMMENTED** [review]
mind share motivation of this addition since it is just one-to-one mapping to trainer attribute...?

File: tests/models/test_restore.py (hunk 4 above)
Comment (nrupatunga) [reply tests/models/test_restore.py:20]: Basically, before the fix, if we resume the training with a particular checkpoint (epoch>1), the following assertions fail. 
`assert trainer.global_step == pl_module.global_step`
`assert trainer.current_epoch == pl_module.current_epoch`

This is because `pl_module.global_step` or `pl_module.current_epoch` properties are not the same as that of `trainer` at different stages. To make sure the same, the test function was added by Adrian.

Comment (ydcjeff) [reply tests/models/test_restore.py]: This probably comes from `self.increment_accumulated_grad_global_step()` of training_loop.py.

File: tests/models/test_restore.py [hunk 5]
```diff
@@ -10,11 +10,52 @@
...
+        self.check_properties(trainer, pl_module)
+
+    def on_train_batch_end(self, trainer, pl_module, *args, **kwargs):
+        self.check_properties(trainer, pl_module)
+
+    def on_epoch_end(self, trainer, pl_module):
+        self.check_properties(trainer, pl_module)
+
+    def on_train_end(self, trainer, pl_module):
+        self.check_properties(trainer, pl_module)
```
Comment (awaelchli) [inline tests/models/test_restore.py:37]:
```suggestion
    def on_train_start(self, trainer, pl_module):
        self._check_properties(trainer, pl_module)

    def on_train_batch_start(self, trainer, pl_module, *args, **kwargs):
        self._check_properties(trainer, pl_module)

    def on_train_batch_end(self, trainer, pl_module, *args, **kwargs):
        self._check_properties(trainer, pl_module)

    def on_epoch_end(self, trainer, pl_module):
        self._check_properties(trainer, pl_module)

    def on_train_end(self, trainer, pl_module):
        self._check_properties(trainer, pl_module)
```

Comment (awaelchli) [reply tests/models/test_restore.py]: I must have renamed it last minute and forgot to apply the changes everywhere... I was pretty tired sorry haha

**Review by nrupatunga — COMMENTED** [review]
Sorry! missed that entirely

File: tests/models/test_restore.py (hunk 4 above)
Comment (awaelchli) [reply tests/models/test_restore.py]: @ydcjeff are you saying there is another bug?
the bug here is simply that when resuming from checkpoint, the model.global_step get's set only in the training loop. This is too late, since you may need it in configure optimizer or other hook that runs right after model is init. I think it make sense to have model grab the global step directly from trainer instead of keeping a separate count. 
Also, by introducing the property here, we make sure the user does not accidentally overrwrite the global step or current epoch in their LightningModule, because if they don't know about it they might try to track their own variables and this would overwrite ours. This is my main concern actually also with the other properties, they are trainer properties duplicated into the LightningModule but there is not guarantee for parity throughout the training loop.

Comment (ydcjeff) [reply tests/models/test_restore.py]: ahhh no no I don't mean like that.
sorry @awaelchli if i confuse you 😓, i just point out where trainer.global_step get incremented.
I will be careful with my comment.

Comment (awaelchli) [reply tests/models/test_restore.py]: @ydcjeff everything is good! don't worry. You don't need to be careful about anything. Maybe my text above was a bit long and intimidating. But the intention was the opposite. I like to explain things in order to understand and give others the chance to disagree with me so I can figure it out. Basically my question was: 
Does the above logic make sense or do we have another bug connected to accumulate gradients?

Comment (ydcjeff) [reply tests/models/test_restore.py]: Yes, it makes sense.

**Comment by nrupatunga** [comment]
Dear @Borda @ydcjeff @awaelchli @ananthsub 

Do we need any other changes or tests on this PR. ?

**Comment by ydcjeff** [comment]
> Dear @Borda @ydcjeff @awaelchli @ananthsub
> 
> Do we need any other changes or tests on this PR. ?

PR is looking good but tests are failing on conda PyTorch 1.7
Can you take a look?

**Comment by ananthsub** [comment]
> PR is looking good but tests are failing on conda PyTorch 1.7 Can you take a look?

The new property decorators are being picked up by JIT scripting and the methods aren't scriptable, so they need to be added to the torch jit ignore list: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/lightning.py#L65-L72

**Comment by nrupatunga** [comment]
> > PR is looking good but tests are failing on conda PyTorch 1.7 Can you take a look?
> 
> The new property decorators are being picked up by JIT scripting and the methods aren't scriptable, so they need to be added to the torch jit ignore list: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/lightning.py#L65-L72

Thank you @ydcjeff and @ananthsub for pointing the solution. I have made the changes and all checks are passed. 

@awaelchli @Borda Hope you guys can approve.

**Comment by ydcjeff** [comment]
@nrupatunga LGTM.
Btw, we need to add this to changelog's `Changed` section.

**Comment by nrupatunga** [comment]
> @nrupatunga LGTM.
> Btw, we need to add this to changelog's `Changed` section.

Sure @ydcjeff, I will add. I should add under **Fixed** section, under **[unreleased] - YYYY-MM-DD**. Sorry first-timer, do help!

**Comment by ydcjeff** [comment]
Yes, this should be under [unreleased] - YYYY-MM-DD.
Yes, **Fixed** is more suitable.

**Comment by nrupatunga** [comment]
> EDIT:
> Shall we mention these changes in the docs too?
> https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#current-epoch
> https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#global-step

@ydcjeff I think this is not needed, as our current changes do not affect it. Isn't it?. As here in the links above, we are just accessing the properties.

**Comment by ydcjeff** [comment]
Yes, this doesn't affect users.
But I think if we already mention in the docs, other users or newcomers can know this they are the same and don't need to ask questions or try comparing them.

What do you think?

**Comment by nrupatunga** [comment]
> Yes, this doesn't affect users.
> But I think if we mention in the docs, other users or newcomers can know this they are the same and don't need to ask questions or try comparing them.
> 
> What do you think?

@ydcjeff 

Should I mention that `self.current_epoch` will be same as `trainer.current_epoch`. (similarly for `global_step`)

I was thinking if that information is really needed for the user to know? As the user will care only about if he could access those at any point in time and expect it to be right.

**Comment by ydcjeff** [comment]
I am suggesting that I prefer to have them in the docs. But, core contributors will decide to add it or not.

**Comment by nrupatunga** [comment]
> I am suggesting that I prefer to have them in the docs. But, core contributors will decide to add it or not.

I agree, I think adding has one advantage that the user will have clarity that `self.current_epoch = trainer.current_epoch`

**Comment by nrupatunga** [comment]
@ydcjeff @ananthsub 
Also, now adding change logs resulted in 4 failed check, Can you help me on why this might be happening

one the logs:
`Failed to fetch https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64/Packages.gz  File has unexpected size (47871 != 49498). Mirror sync in progress? [IP: 152.195.19.142 443]`

Is there a way to rerun them

**Comment by ydcjeff** [comment]
Don't worry about the failed ones, these are unrelated to this PR.

File: CHANGELOG.md [hunk 6]
```diff
@@ -93,6 +93,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed aggregation of metrics ([#3517](https://github.com/PyTorchLightning/pytorch-lightning/pull/3517))
 
+- Fixed `current_epoch` and `global_step` properties mismatch between `trainer` and lightning `module` ([#3785](https://github.com/PyTorchLightning/pytorch-lightning/pull/3785))
```
Comment (awaelchli) [inline CHANGELOG.md:96]:
```suggestion
- Fixed `current_epoch` and `global_step` properties mismatch between `Trainer` and `LightningModule` ([#3785](https://github.com/PyTorchLightning/pytorch-lightning/pull/3785))
```

**Review by awaelchli — APPROVED** [review]
👍 nice

**Comment by Borda** [comment]
Seems that some tests are failing...

File: tests/models/test_restore.py [hunk 7]
```diff
@@ -10,11 +10,52 @@
...
+
+    def on_train_end(self, trainer, pl_module):
+        self._check_properties(trainer, pl_module)
+
+
+def test_resume_from_checkpoint(tmpdir):
+    """ Test that properties like `current_epoch` and `global_step`
+    in model and trainer are always the same. """
+    model = EvalModelTemplate()
+    checkpoint_callback = ModelCheckpoint(filepath=tmpdir, monitor="val_loss", save_last=True)
```
Comment (awaelchli) [inline tests/models/test_restore.py:44]:
```suggestion
    checkpoint_callback = ModelCheckpoint(filepath=tmpdir, monitor="early_stop_on", save_last=True)
```

Comment (awaelchli) [reply tests/models/test_restore.py:44]: > Seems that some tests are failing...

because of changes in master

File: tests/models/test_restore.py [hunk 8]
```diff
@@ -10,11 +10,52 @@
...
+    def on_epoch_end(self, trainer, pl_module):
+        self._check_properties(trainer, pl_module)
+
+    def on_train_end(self, trainer, pl_module):
+        self._check_properties(trainer, pl_module)
+
+
+def test_resume_from_checkpoint(tmpdir):
+    """ Test that properties like `current_epoch` and `global_step`
+    in model and trainer are always the same. """
```
Comment (rohitgr7) [inline tests/models/test_restore.py:42]: I think we should extend this test somehow to check for all the things that are required to be restored. Not here maybe.

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
